packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -381,13 +401,33 @@ export class TokenDetectionController extends StaticIntervalPollingController<
     }, this.getIntervalLength());
   }
 
-  #getCorrectChainId(networkClientId?: NetworkClientId) {
-    const { chainId } =
-      this.messagingSystem.call(
+  #getCorrectChainIdAndNetworkClientId(networkClientId?: NetworkClientId): {
+    chainId: Hex;
+    networkClientId: NetworkClientId;
+  } {
+    if (networkClientId) {
+      const networkConfiguration = this.messagingSystem.call(
         'NetworkController:getNetworkConfigurationByNetworkClientId',
-        networkClientId ?? this.#networkClientId,
-      ) ?? {};
-    return chainId ?? this.#chainId;
+        networkClientId,
+      );
+      if (networkConfiguration) {
+        return {
+          chainId: networkConfiguration.chainId,
+          networkClientId,
+        };
+      }
+    }
+    const { chainId } = this.messagingSystem.call(
+      'NetworkController:getProviderConfig',
+    );
+    const newNetworkClientId = this.messagingSystem.call(
+      'NetworkController:findNetworkClientIdByChainId',
+      this.#chainId,
+    );
+    return {
+      chainId,
+      networkClientId: newNetworkClientId,
+    };
```

Comment:
Applies new logic for retrieving `networkClientId` introduced [here](https://github.com/MetaMask/metamask-extension/pull/22898/files#diff-323d0cf46498be3850b971474905354ea5ccf7fa13745ad1e6eba59c5b586830R131-R143). 
- Uses `getNetworkConfigurationByNetworkClientId` action instead of `getNetworkClientById`. 
- If `networkConfiguration` returns as `undefined`, falls back to using `getProviderConfig` and `findNetworkClientIdByChainId` actions.

Reply:
looking at this logic again, i'm unsure we we don't just call `NetworkController:getState` and read off `providerConfig.chainId` and `selectedNetworkClientId`

Reply:
Looks like `findNetworkClientIdByChainId` returns the first network client in `getNetworkClientRegistry()` with matching `chainId`. If that's always consistent with `selectedNetworkClientId` then using `getState` shouldn't create any differences? 

I guess more granularly scoped actions/events might make tests easier to mock/write, and potentially help with security through encapsulation.